### ISSUE TITLE
feat: Allow to specify in CustomOuputTarget if the output have to be re-executed on each build

### ISSUE THREAD

**Issue by Sukaato** [ionic-team/stencil#5514]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already include this feature request, without success.

### Describe the Feature Request

I'd like to have a parameter that lets you choose whether the custom output target can be called at every change or only during the "final" build.

### Describe the Use Case

I've created a little plugin to retrieve all `@config` docsTags in a `component-config.d.ts` so it's easyer to customise the default state of a component.

stencil.config.ts
```ts
export const config: Config = {
  ... // Config
  outputTargets: [
    {
      type: 'custom',
      name: 'component-option',
     execType: 'run', // new prop in CustomOutputTarget
     async generator(config, compilerCtx: any, buildCtx, docs): Promise<void> {
       // Logique to create a .d.ts file
     }
    },
  ]
}


```
component-config.d.ts
```ts
export interface ComponentOptions {
  "pop-drawer"?: {
    triggerAction?: "click" | "context-menu" | "hover";
    side?: "end" | "start";
    open?: boolean;
  }
}

```
These options are extracted from the component, but if a modification is made to this component, the output target is not updated.

### Describe Preferred Solution

```ts
export const config: Config = {
  ... // Config
  outputTargets: [
    {
      type: 'custom',
      name: 'component-option',
     execType: 'run' | 'build', // new prop in CustomOutputTarget
     async generator(config, compilerCtx: any, buildCtx, docs): Promise<void> { }
    },
  ]
}

// 'run' : run on every change
// 'build' : run only on "final" build
```

### Related Code

https://github.com/Sukaato/stencil-custom-target-example

**Comment by christian-bromann** [comment]
> I'd like to have a parameter that lets you choose whether the custom output target can be called at every change or only during the "final" build.

Can you elaborate on this? When you say "at every change" are you talking about having the output target run during watch mode? Because this works for me given your example project. I guess, could you explain in detail what the current behavior is and your expected/desired behavior?

**Comment by Sukaato** [comment]
Yes when a mean at every change, i mean in dev mode.

Like i said, when i add or remove a `@config` in JSdoc, the `component-config.d.ts` is not updated (not the components.d.ts that stencil provide),

I add some console.log in my component-config.target.ts but there only called on final build, not in dev mode.

**Comment by christian-bromann** [comment]
Gotcha, what I can observe is that the `src/components.d.ts` is changing after every save:

![drawer](https://github.com/ionic-team/stencil/assets/731337/7892eca2-8bbf-4399-9571-de015f0b01c0)

However it seems that VSCode is not picking this up:
<img width="1163" alt="Screenshot 2024-03-19 at 2 56 44 PM" src="https://github.com/ionic-team/stencil/assets/731337/f4140467-5268-4bd9-9442-04f26df80e55">

I will ingest this into our backlog for the team to take a closer look at. Any contributions are welcome!

**Comment by Sukaato** [comment]
After running a few tests, I realized that the custom output target is never called when the `--dev` flag is set in the CLI, when it is cleared, the output target is called on first compilation, and on rebuild.

Based on this result, I think the bug is related to the dev flag.

**Comment by alicewriteswrongs** [comment]
This was closed in #5541, and that fix will be released today!

**Comment by alicewriteswrongs** [comment]
The fix for this issue has been released as a part of today's [Stencil v4.18.1 release](https://github.com/ionic-team/stencil/releases/tag/v4.18.1)!
